File: katas/python/OOP/polymorphism/polymorphism_bad.py
# Violación de Polimorfismo: Lógica condicional para tipos
# ❌ Problema: Usar if/else o switch para diferentes tipos

from typing import List, Literal, TypedDict


# ❌ Sin herencia ni polimorfismo - solo diccionarios
class DogData(TypedDict):
    type: Literal["dog"]
    name: str


class CatData(TypedDict):
    type: Literal["cat"]
    name: str


class BirdData(TypedDict):
    type: Literal["bird"]
    name: str


AnimalData = DogData | CatData | BirdData


# ❌ Clase que necesita conocer todos los tipos específicos
class AnimalProcessor:
    # ❌ Método lleno de condicionales para cada tipo
    def make_sound(self, animal: AnimalData) -> None:
        if animal["type"] == "dog":
            print(f"{animal['name']} dice: ¡Guau guau!")
        elif animal["type"] == "cat":
            print(f"{animal['name']} dice: ¡Miau miau!")
        elif animal["type"] == "bird":
            print(f"{animal['name']} dice: ¡Pío pío!")
        # ❌ Si agrego un nuevo animal, debo modificar este método

    # ❌ Más condicionales para cada comportamiento
    def feed(self, animal: AnimalData) -> None:
        if animal["type"] == "dog":
            print(f"{animal['name']} está comiendo croquetas")
        elif animal["type"] == "cat":
            print(f"{animal['name']} está comiendo pescado")
        elif animal["type"] == "bird":
            print(f"{animal['name']} está comiendo semillas")

    # ❌ Y más condicionales para movimiento
    def move(self, animal: AnimalData) -> None:
        if animal["type"] == "dog":
            print(f"{animal['name']} está corriendo")
        elif animal["type"] == "cat":
            print(f"{animal['name']} está saltando")
        elif animal["type"] == "bird":
            print(f"{animal['name']} está volando")


if __name__ == "__main__":
    # ❌ Uso con condicionales por todos lados
    print("=== Violación de Polimorfismo ===")

    processor = AnimalProcessor()

    dog: DogData = {"type": "dog", "name": "Rex"}
    cat: CatData = {"type": "cat", "name": "Luna"}
    bird: BirdData = {"type": "bird", "name": "Piolín"}

    animals: List[AnimalData] = [dog, cat, bird]

    # ❌ El procesador debe verificar el tipo constantemente
    for animal in animals:
        processor.make_sound(animal)
        processor.feed(animal)
        processor.move(animal)

    # ❌ Problemas:
    # - Muchos if/else y switch repetidos
    # - Agregar nuevo animal requiere modificar MUCHOS métodos
    # - Propenso a errores (olvidar un caso)
    # - Viola Open/Closed Principle
    # - Difícil de mantener y escalar

File: katas/php/OOP/polymorphism/polymorphism-bad.php
<?php
// Violación de Polimorfismo: Lógica condicional para tipos
// ❌ Problema: Usar if/else o switch para diferentes tipos

// ❌ Sin herencia ni polimorfismo - solo arrays asociativos
class AnimalProcessor {
    // ❌ Método lleno de condicionales para cada tipo
    public function makeSound(array $animal): void {
        if ($animal['type'] === 'dog') {
            echo "{$animal['name']} dice: ¡Guau guau!" . PHP_EOL;
        } elseif ($animal['type'] === 'cat') {
            echo "{$animal['name']} dice: ¡Miau miau!" . PHP_EOL;
        } elseif ($animal['type'] === 'bird') {
            echo "{$animal['name']} dice: ¡Pío pío!" . PHP_EOL;
        }
        // ❌ Si agrego un nuevo animal, debo modificar este método
    }

    // ❌ Más condicionales para cada comportamiento
    public function feed(array $animal): void {
        if ($animal['type'] === 'dog') {
            echo "{$animal['name']} está comiendo croquetas" . PHP_EOL;
        } elseif ($animal['type'] === 'cat') {
            echo "{$animal['name']} está comiendo pescado" . PHP_EOL;
        } elseif ($animal['type'] === 'bird') {
            echo "{$animal['name']} está comiendo semillas" . PHP_EOL;
        }
    }

    // ❌ Y más condicionales para movimiento
    public function move(array $animal): void {
        if ($animal['type'] === 'dog') {
            echo "{$animal['name']} está corriendo" . PHP_EOL;
        } elseif ($animal['type'] === 'cat') {
            echo "{$animal['name']} está saltando" . PHP_EOL;
        } elseif ($animal['type'] === 'bird') {
            echo "{$animal['name']} está volando" . PHP_EOL;
        }
    }
}

// ❌ Uso con condicionales por todos lados
echo "=== Violación de Polimorfismo ===" . PHP_EOL;

$processor = new AnimalProcessor();

$dog = ['type' => 'dog', 'name' => 'Rex'];
$cat = ['type' => 'cat', 'name' => 'Luna'];
$bird = ['type' => 'bird', 'name' => 'Piolín'];

$animals = [$dog, $cat, $bird];

// ❌ El procesador debe verificar el tipo constantemente
foreach ($animals as $animal) {
    $processor->makeSound($animal);
    $processor->feed($animal);
    $processor->move($animal);
}

// ❌ Problemas:
// - Muchos if/else y switch repetidos
// - Agregar nuevo animal requiere modificar MUCHOS métodos
// - Propenso a errores (olvidar un caso)
// - Viola Open/Closed Principle
// - Difícil de mantener y escalar
